src/server/proxy_protocol.rs

```diff
@@ -159,71 +214,88 @@ where
     U: UserDetail + 'static,
 {
     pub fn new(logger: slog::Logger, passive_ports: Range<u16>) -> Self {
-        let board = HashMap::new();
+        let board = DashMap::new();
         Self {
             switchboard: board,
             port_range: passive_ports,
             logger,
         }
     }
 
-    fn try_and_claim(&mut self, hash: String, session_arc: SharedSession<S, U>) -> Result<(), ProxyProtocolError> {
-        match self.switchboard.get(&hash) {
-            Some(_) => Err(ProxyProtocolError::EntryNotAvailable),
-            None => match self.switchboard.insert(hash, Some(session_arc)) {
-                Some(_) => {
-                    slog::warn!(self.logger, "This is a data race condition. This shouldn't happen");
-                    // just return Ok anyway however
-                    Ok(())
-                }
-                None => Ok(()),
-            },
+    pub async fn try_and_claim(&mut self, hash: &ProxyHashKey, session_arc: SharedSession<S, U>) -> Result<(), ProxyProtocolError> {
+        // Atomically insert the key and value into the switchboard hashmap
+        match self.switchboard.entry(hash.to_string()) {
```

Comment:
Why do we need `.to_string()`? The string representation is probably bigger than the original struct is, and we already derive hash on the struct, so why not just store the struct directly to avoid string conversion?

Reply:
The hash by itself isn't hashable (some trait(s) need to be implemented), which is why I resorted to string conversion temporarily. I wanted to get it functionally working first and not be bothered too much with these details. I'll see if I can make this work without the string.